pkg/proxy/endpoints.go

```diff
@@ -0,0 +1,231 @@
+	// items is a service to its endpointsChange map.
+	items map[types.NamespacedName]*endpointsChange
+}
+
+// endpointsChange contains all changes to endpoints that happened since proxy rules was synced.  For a single object,
+// changes are accumulated, i.e. previous is state from before all of them,
+// current is state after applying all of those.
+type endpointsChange struct {
+	previous EndpointsMap
+	current  EndpointsMap
+}
+
+// EndpointsMap maps a service to its endpoints.
+type EndpointsMap map[ServicePortName][]Endpoints
+
+// NewEndpointsChangeMap initialize an EndpointsChangeMap
+func NewEndpointsChangeMap(hostname string) EndpointsChangeMap {
+	return EndpointsChangeMap{
+		hostname: hostname,
+		items:    make(map[types.NamespacedName]*endpointsChange),
+	}
+}
+
+// UpdateEndpointMapResult is the updated results after applying endpoints changes.
+type UpdateEndpointMapResult struct {
+	// HCEndpointsLocalIPSize maps an endpoints name to the length of its local IPs.
+	HCEndpointsLocalIPSize map[types.NamespacedName]int
+	// StaleServiceNames identifies if an endpoints service pair is stale.
+	StaleEndpoints map[EndpointServicePair]bool
+	// StaleServiceNames identifies if a service is stale.
+	StaleServiceNames map[ServicePortName]bool
+}
+
+// UpdateEndpointsMap updates endpointsMap base on the given changes.
+// The changes map is cleared after applying them.
+func UpdateEndpointsMap(endpointsMap EndpointsMap, changes *EndpointsChangeMap, hostname string) (result UpdateEndpointMapResult) {
+	result.StaleEndpoints = make(map[EndpointServicePair]bool)
+	result.StaleServiceNames = make(map[ServicePortName]bool)
+
+	func() {
+		endpointsMap.apply(changes, result.StaleEndpoints, result.StaleServiceNames)
+	}()
+
+	// TODO: If this will appear to be computationally expensive, consider
+	// computing this incrementally similarly to endpointsMap.
+	result.HCEndpointsLocalIPSize = make(map[types.NamespacedName]int)
+	localIPs := GetLocalIPs(endpointsMap)
+	for nsn, ips := range localIPs {
+		result.HCEndpointsLocalIPSize[nsn] = len(ips)
+	}
+
+	return result
+}
+
+// apply the changes to EndpointsMap and updates stale endpoints and service-endpoints pair. The `staleEndpoints` argument
+// is passed in to store the stale udp endpoints and `staleServiceNames` argument is passed in to store the stale udp service.
+func (endpointsMap EndpointsMap) apply(changes *EndpointsChangeMap, staleEndpoints map[EndpointServicePair]bool, staleServiceNames map[ServicePortName]bool) {
+	changes.lock.Lock()
+	defer changes.lock.Unlock()
+	for _, change := range changes.items {
+		endpointsMap.Unmerge(change.previous)
+		endpointsMap.Merge(change.current)
+		detectStaleConnections(change.previous, change.current, staleEndpoints, staleServiceNames)
+	}
+	changes.items = make(map[types.NamespacedName]*endpointsChange)
+}
+
+// Merge adds other ProxyEndpointsMap's elements to current ProxyEndpointsMap, if exists, update it, otherwise create new one.
```

Comment:
I think a much more concise comment for this would read as follows: "Merge ensures that the current EndpointsMap for the proxier is rebased onto the passed in EndpointsMap" WDYT?

Reply:
Nice comment, will follow it.

Reply:
Done.